proposals/module-types/Explainer.md

```diff
@@ -0,0 +1,745 @@
+  (func (export "run")
+    call $i.$hi
+  )
+)
+```
+
+Unlike most source-language nested functions/classes, nested modules have no
+special access to their parents' state. However, since modules and types are
+closed, stateless expressions and thus "sharing" is equivalent to duplication,
+nested modules can refer to their parents' types and modules via *aliases*. As
+syntactic sugar in the textual format, a nested module can simply use the
+`$name` of the enclosing module/type:
+```wasm
+(module
+  (type $WasiFile (instance $wasi-file
+    (export "read" (func (param i32 i32 i32) (result i32)))
+  ))
+  (module $child
+    (import "wasi_file" (instance (type $WasiFile)))
+  )
+)
+```
+The desugared form of `$child` would explicitly create the alias with a `type`
+definition which references the parent:
+```wasm
+(module $Parent
+  (type $WasiFile (instance $wasi-file
+    (export "read" (func (param i32 i32 i32) (result i32)))
+  ))
+  (module $child
+    (type $WasiFile (alias $Parent.$WasiFile))
+    (import "wasi_file" (instance (type $WasiFile)))
+  )
+)
+```
+In the binary format, a type alias could be represented by a [De Bruijn index]
+relative to the sequence of parent modules and their respective module/type
+index spaces.
+
+Other than type/module sharing, which can easily be undone by duplication, a
+module A that imports module B is semantically equivalent to module A nesting
+module B. This means tools can merge multiple `.wasm` files into one `.wasm` by
+a trivial, language-independent transformation. Thus, module imports and nested
+modules can serve as a language-independent basis for first expressing a module
+dependency graph and then bundling/linking that dependency graph into a single
+`.wasm` module for portable distribution.
+
+Additionally, on the Web and other hosts that provide runtime instantiation
+APIs, nested modules allow an outer module to bundle module code that is to be
+instantiated dynamically by passing a first-class module reference (created via
+`ref.module`) to the runtime instantiation API.
+
+
+### Module and Instance Exports
+
+Lastly, symmetric to all other kinds of definitions, modules and instances can
+be exported. For example:
+```wasm
+(module
+  (import "a" (module $a ...))
+  (module $b ...)
+  (import "c" (instance $c ...))
+  (instance $d ...)
+
+  (export "e1" (module $a))
+  (export "e2" (module $b))
+  (export "e3" (instance $c))
+  (export "e4" (instance $d))
+)
+```
+Therefore, module and instance types can appear in both the imports and exports
+of module types and instance types.
+
+
+### Summary
+
+To summarize the proposed changes (all changes in both text and binary format):
+* The `module` field of [`import`] become optional (allowing single-level
+  imports). (How to encode this in the [import section] is an interesting
+  question.)
+* New `module` and `instance` type constructors are added that may be used to
+  define types in the [type section].
```

Comment:
As mentioned above, I think we'll also need `global`, `memory`, and `table` entries in the type section, otherwise we can't type `ref.table` and friends.

Reply:
Thanks, fixed.